Mage/Archivist/Archivist.cs

```diff
@@ -78,36 +78,31 @@ public class Archivist : EventEmitter<VaultValue> {
 
 	// Returns cache value if it exists and has not passed max allowed age
 	private VaultValue GetCacheValue(string cacheKeyName, int? maxAge = null) {
-		if (!_cache.ContainsKey(cacheKeyName)) {
-			return null;
-		}
+		lock ((object)this) {
+			if (!_cache.ContainsKey(cacheKeyName)) {
+				return null;
+			}
 
-		VaultValue value = _cache[cacheKeyName];
-		double timespan = (DateTime.Now - value.writtenAt).TotalMilliseconds;
-		if (maxAge != null && timespan > maxAge * 1000) {
-			return null;
-		}
+			VaultValue value = _cache[cacheKeyName];
+			double timespan = (DateTime.Now - value.writtenAt).TotalMilliseconds;
+			if (maxAge != null && timespan > maxAge * 1000) {
+				return null;
+			}
 
-		return value;
+			return value;
+		}
 	}
 	
 
 	// Ensures a cache value exists then sets it based on info provided
-	private void SetCacheValue(JObject info) {
-		string topic = (string)info["key"]["topic"];
-		JObject index = (JObject)info["key"]["index"];
-		string cacheKeyName = GetCacheKey(topic, index);
-		VaultValue cacheValue = GetCacheValue(cacheKeyName);
-		if (cacheValue == null) {
-			cacheValue = new VaultValue(topic, index);
+	private VaultValue CreateCacheValue(string topic, JObject index) {
+		lock ((object)this) {
+			VaultValue cacheValue = new VaultValue(topic, index);
+
+			string cacheKeyName = GetCacheKey(topic, index);
```

Comment:
GetCacheKey never uses "this" object (you may wanna make it static because of that and I would argue `createCacheKey` would be more intuitive a name, but that's outside this PR), so there's no point in locking that. The same seems to be true for `new VaultValue`. So you only need to lock the _cache.Add call.